elixir/apps/domain/lib/domain/okta/sync.ex

```diff
@@ -0,0 +1,815 @@
+        sync(directory)
+    end
+
+    :ok
+  end
+
+  def perform(_), do: :ok
+
+  defp update(directory, attrs) do
+    changeset =
+      Ecto.Changeset.cast(directory, attrs, [
+        :synced_at,
+        :error_email_count,
+        :error_message,
+        :errored_at,
+        :is_disabled,
+        :disabled_reason,
+        :is_verified
+      ])
+
+    {:ok, _directory} = changeset |> Safe.unscoped() |> Safe.update()
+  end
+
+  defp sync(%Okta.Directory{} = directory) do
+    client = Okta.APIClient.new(directory)
+    access_token = get_access_token!(client, directory)
+    synced_at = DateTime.utc_now()
+
+    apps = get_apps!(client, access_token, directory)
+    sync_all_apps!(apps, client, access_token, directory, synced_at)
+    sync_all_memberships!(client, access_token, directory, synced_at)
+    delete_unsynced(directory, synced_at)
+
+    # Clear error state on successful sync completion
+    update(directory, %{
+      "synced_at" => synced_at,
+      "error_email_count" => 0,
+      "error_message" => nil,
+      "errored_at" => nil,
+      "is_disabled" => false,
+      "disabled_reason" => nil
+    })
+
+    duration = DateTime.diff(DateTime.utc_now(), synced_at)
+
+    Logger.info("Finished Okta directory sync in #{duration} seconds",
+      okta_directory_id: directory.id
+    )
+  end
+
+  defp get_access_token!(client, directory) do
+    Logger.debug("Getting access token", okta_directory_id: directory.id)
+
+    case Okta.APIClient.fetch_access_token(client) do
+      {:ok, access_token} ->
+        Logger.debug("Successfully obtained access token", okta_directory_id: directory.id)
+        access_token
+
+      {:error, error} ->
+        Logger.debug("Failed to get access token",
+          okta_directory_id: directory.id,
+          error: inspect(error)
+        )
+
+        raise Okta.SyncError,
+          reason: "Failed to get access token",
+          cause: error,
+          directory_id: directory.id,
+          step: :get_access_token
+    end
+  end
+
+  defp get_apps!(client, token, directory) do
+    Logger.debug("Fetching Okta apps", okta_directory_id: directory.id)
+
+    case Okta.APIClient.list_apps(client, token) do
+      {:ok, apps} ->
+        Logger.debug("Successfully fetched Okta apps",
+          okta_directory_id: directory.id,
+          count: length(apps)
+        )
+
+        apps
+
+      {:error, error} ->
+        Logger.debug("Failed to fetch Okta apps",
+          okta_directory_id: directory.id,
+          error: inspect(error)
+        )
+
+        raise Okta.SyncError,
+          reason: "Failed to fetch apps",
+          cause: error,
+          directory_id: directory.id,
+          step: :list_apps
+    end
+  end
+
+  # Process all apps, syncing users and groups for each app
+  defp sync_all_apps!(apps, client, token, directory, synced_at) do
+    Enum.each(apps, fn app ->
+      sync_single_app!(app, client, token, directory, synced_at)
+    end)
+  end
```

Comment:
Don't we only sync one app per sync run / directory?

Reply:
I may not be following - we already schedule each directory to sync right? In the scheduler worker. That should insert N jobs for all Okta directories in the DB.

Then, when one runs, you're saying there's multiple apps tied to a single okta_domain that we need to fetch and sync individually?

Reply:
I'm probably confused - in Entra there's the one `Firezone Authentication` app the admin will assign users/groups to, and we find that single one, grab the assignments and go to town. I guess Okta's different?